metricflow/time/time_granularity_solver.py

```diff
@@ -116,35 +128,75 @@ def resolve_granularity_for_partial_time_dimension_specs(
 
-        if not minimum_time_granularity:
-            raise RequestTimeGranularityException(
-                f"Unable to resolve the time dimension spec for {partial_time_dimension_spec}. "
-                f"Valid group by elements are:\n"
-                f"{pformat_big_objects([spec.qualified_name for spec in valid_group_by_elements.as_spec_set.as_tuple])}"
+        if metric_references:
+            valid_group_by_elements = self._semantic_manifest_lookup.metric_lookup.linkable_set_for_metrics(
+                metric_references=metric_references,
             )
+            for path_key in valid_group_by_elements.path_key_to_linkable_dimensions:
+                if (
+                    path_key.element_name == partial_time_dimension_spec.element_name
+                    and path_key.entity_links == partial_time_dimension_spec.entity_links
+                    and path_key.time_granularity is not None
+                ):
+                    minimum_time_granularity = (
+                        path_key.time_granularity
+                        if minimum_time_granularity is None
+                        else min(minimum_time_granularity, path_key.time_granularity)
+                    )
+            if not minimum_time_granularity:
+                raise RequestTimeGranularityException(
+                    f"Unable to resolve the time dimension spec for {partial_time_dimension_spec}. "
+                    f"Valid group by elements are:\n"
+                    f"{pformat_big_objects([spec.qualified_name for spec in valid_group_by_elements.as_spec_set.as_tuple])}"
+                )
+        else:
+            minimum_time_granularity = self.get_min_granularity_for_partial_time_dimension_without_metrics(
+                read_nodes=read_nodes,
+                node_output_resolver=node_output_resolver,
+                partial_time_dimension_spec=partial_time_dimension_spec,
+            )
+            if not minimum_time_granularity:
+                raise RequestTimeGranularityException(
+                    f"Unable to resolve the time dimension spec for {partial_time_dimension_spec}. "
+                )
 
         return minimum_time_granularity
 
+    def get_min_granularity_for_partial_time_dimension_without_metrics(
+        self,
+        read_nodes: Sequence[BaseOutput],
+        node_output_resolver: DataflowPlanNodeOutputDataSetResolver,
+        partial_time_dimension_spec: PartialTimeDimensionSpec,
+    ) -> Optional[TimeGranularity]:
+        """Find the minimum."""
+        granularity_free_qualified_name = StructuredLinkableSpecName(
+            entity_link_names=tuple(
+                [entity_link.element_name for entity_link in partial_time_dimension_spec.entity_links]
+            ),
+            element_name=partial_time_dimension_spec.element_name,
+        ).granularity_free_qualified_name
+        for read_node in read_nodes:
+            output_data_set = node_output_resolver.get_output_data_set(read_node)
```

Comment:
This feels like potentially a lot of iterating through datasets to me since we'll call this function once for every partial time dimension spec, but I think it's fine for now.

Reply:
Yeah, it would probably be better to create a dict of time dimension instances to their output datasets on init (or something like that) so we only need to do this loop once. Not sure how complicated that change will be, let me look into it and see if it's worth doing before merge

Reply:
FWIW, it turns out we are caching the output dataset so `get_output_data_set()` is just a dictionary lookup after the first call for a given node!